build.rs

```diff
@@ -230,3 +500,203 @@ fn run_command_with_args<S>(command_name: S, args: &Vec<String>)
+            run_command_with_args(&get_command("PERL_EXECUTABLE", "perl"),
+                                  &args);
+        }
+        r
+    } else {
+        p.to_str().expect("Could not convert path").into()
+    }
+}
+
+fn need_run(source: &Path, target: &Path) -> bool {
+    let s = std::fs::metadata(source);
+    let t = std::fs::metadata(target);
+    if s.is_err() || t.is_err() {
+        true
+    } else {
+        match (s.unwrap().modified(), t.unwrap().modified()) {
+            (Ok(s), Ok(t)) => s > t,
+            _ => true,
+        }
+    }
+}
+
+fn get_command(var: &str, default: &str) -> String {
+    env::var(var).unwrap_or(default.into())
+}
+
+
+fn check_all_files_tracked() {
+    walk_dir(&PathBuf::from("crypto"), &is_tracked);
+    walk_dir(&PathBuf::from("include"), &is_tracked);
+    walk_dir(&PathBuf::from("mk"), &is_tracked);
+}
+
+
+fn is_tracked(file: &DirEntry) {
+    let p = file.path();
+    let cmp = |f| p == PathBuf::from(f);
+    let tracked = match p.extension().and_then(|p| p.to_str()) {
+        Some("h") => {
+            RING_HEADERS.iter().chain(RING_TEST_HEADERS.iter()).any(cmp)
+        },
+        Some("inl") => RING_INLINE_FILES.iter().any(cmp),
+        Some("mk") => RING_MAKEFILES.iter().any(cmp),
+        Some("c") | Some("cc") => {
+            RING_SRC.iter()
+                .chain(RING_AARCH64_SRCS.iter())
+                .chain(RING_ARM_SHARED_SRCS.iter())
+                .chain(RING_ARM_SRCS.iter())
+                .chain(RING_INTEL_SHARED_SRCS.iter())
+                .chain(RING_TEST_SRCS.iter())
+                .chain(RING_X86_64_SRC.iter())
+                .chain(RING_X86_SRCS.iter())
+                .any(cmp)
+        },
+        Some("S") => {
+            RING_AARCH64_SRCS.iter()
+                .chain(RING_ARM_SHARED_SRCS.iter())
+                .chain(RING_ARM_SRCS.iter())
+                .chain(RING_INTEL_SHARED_SRCS.iter())
+                .chain(RING_X86_64_SRC.iter())
+                .chain(RING_X86_SRCS.iter())
+                .any(cmp)
+        },
+        Some("pl") => {
+            RING_AARCH64_SRCS.iter()
+                .chain(RING_ARM_SHARED_SRCS.iter())
+                .chain(RING_ARM_SRCS.iter())
+                .chain(RING_INTEL_SHARED_SRCS.iter())
+                .chain(RING_X86_64_SRC.iter())
+                .chain(RING_X86_SRCS.iter())
+                .chain(RING_PERL_INCLUDES.iter())
+                .any(cmp)
+        },
+        _ => true,
+    };
+    if !tracked {
+        panic!("{:?} is not tracked in build.rs", p);
+    }
+}
+
+fn walk_dir<F>(dir: &Path, cb: &F)
+    where F: Fn(&DirEntry)
+{
+    if dir.is_dir() {
+        for entry in fs::read_dir(dir).unwrap() {
+            let entry = entry.unwrap();
+            let path = entry.path();
+            if path.is_dir() {
+                walk_dir(&path, cb);
+            } else {
+                cb(&entry);
+            }
+        }
+    }
+}
+
+trait BuildHelper {
```

Comment:
What is the purpose of this trait? In particular, srcs() is only called in one place, asmflags() is only called in one place, etc. And there's only one implementation of the trait. So, why not just get rid of the trait and calculate these values within the functions that currently call these methods? That would be much clearer to me because each function would be more self-contained.